snapcraft/commands/upload.py

```diff
@@ -42,20 +45,42 @@
 logger = logging.getLogger(__name__)
 
 
+def _get_name_from_snap_file(snap_path):
+    with tempfile.TemporaryDirectory() as temp_dir:
+        subprocess.check_call(
+            ['unsquashfs', '-d', os.path.join(temp_dir, 'squashfs-root'),
+             snap_path, '-e', os.path.join('meta', 'snap.yaml')])
+        with open(os.path.join(
+                temp_dir, 'squashfs-root', 'meta', 'snap.yaml')) as yaml_file:
+            snap_yaml = yaml.load(yaml_file)
+
+    return snap_yaml['name']
+
+
 def main(argv=None):
     """Upload snap package to the Ubuntu Store."""
     argv = argv if argv else []
-    docopt(__doc__, argv=argv)
+    args = docopt(__doc__, argv=argv)
+
+    if args['SNAP_FILE']:
+        snap_filename = args['SNAP_FILE']
+        if not os.path.exists(snap_filename):
+            raise FileNotFoundError(snap_filename)
+        else:
+            snap_name = _get_name_from_snap_file(snap_filename)
+            logger.info('Uploading existing {}.'.format(snap_filename))
+    else:
+        # make sure the full lifecycle is executed
+        yaml_config = snapcraft.yaml.load_config()
+        snap_filename = format_snap_name(yaml_config.data)
 
-    # make sure the full lifecycle is executed
-    yaml_config = snapcraft.yaml.load_config()
-    snap_filename = format_snap_name(yaml_config.data)
+        if not os.path.exists(snap_filename):
+            logger.info(
+                'Snap {} not found. Running snap step to create it.'.format(
+                    snap_filename))
+            snap.main(argv=argv)
 
-    if not os.path.exists(snap_filename):
-        logger.info(
-            'Snap {} not found. Running snap step to create it.'.format(
-                snap_filename))
-        snap.main(argv=argv)
+        snap_name = yaml_config.data['name']
```

Comment:
I think that's what I'm fixing with Sergio's comment. I'm currently failing if the in-tree yaml is name different than the snap.yaml. Let me know if you would prefer otherwise. A new rev is coming...